core/src/main/java/org/opensearch/sql/analysis/ExpressionReferenceOptimizer.java

```diff
@@ -70,7 +78,30 @@ public Expression visitFunction(FunctionExpression node, AnalysisContext context
       final List<Expression> args =
           node.getArguments().stream().map(expr -> expr.accept(this, context))
               .collect(Collectors.toList());
-      return (Expression) repository.compile(node.getFunctionName(), args);
+      String funcName = node.getFunctionName().toString();
+      FunctionImplementation ret = repository.compile(node.getFunctionName(), args);
+
+      TypeEnvironment typeEnv = context.peek();
+      if (isSingleFieldFunction(funcName)) {
+        ret.getArguments().stream().filter(arg ->
+            (((NamedArgumentExpression)arg).getArgName().equals("field"))
+                && !((NamedArgumentExpression)arg).getValue().toString().contains("*")
+        ).findFirst().ifPresent(arg ->
+            typeEnv.resolve(new Symbol(Namespace.FIELD_NAME,
+                StringUtils.unquoteText(((NamedArgumentExpression)arg).getValue().toString()))
+            )
+        );
+      } else if (isMultiFieldFunction(funcName)) {
+        ret.getArguments().stream().filter(arg ->
+            ((NamedArgumentExpression)arg).getArgName().equals("fields")
+        ).findFirst().ifPresent(fields ->
+            ((NamedArgumentExpression)fields).getValue().valueOf(null).tupleValue()
+                .entrySet().stream().filter(k -> !(k.getKey().contains("*"))
+                ).forEach(key -> typeEnv.resolve(new Symbol(Namespace.FIELD_NAME, key.getKey())))
+        );
+      }
+
+      return (Expression) ret;
```

Comment:
@forestmvey can you combine this block and `isSingleFieldFunction`/`isMultiFieldFunction` into a separate class  derived from `ExpressionNodeVisitor<Expression, AnalysisContext>`? It'd need to be instantiated in the same places `ExpressionReferenceOptimizer`.

Down the road, it will need to work with https://github.com/opensearch-project/sql/issues/811. It'll be easier to do so if it's in a separate class.

Reply:
Wasn't too much of a fan of implementing a visitor that doesn't visit. I moved the validation step to `OpenSearchFunctions` class. Let me know if you think this is suitable.